baybe/surrogates/utils.py

```diff
@@ -24,47 +29,31 @@ def catch_constant_targets(cls: type[Surrogate], std_threshold: float = 1e-6):
 
     The modified class handles the above cases separately from "regular operation"
     by resorting to a :class:`baybe.surrogates.naive.MeanPredictionSurrogate`,
-    which is stored as an additional temporary attribute in its objects.
+    which is stored outside the model in a dictionary maintained by this decorator.
 
     Args:
         cls: The :class:`baybe.surrogates.base.Surrogate` to be augmented.
         std_threshold: The standard deviation threshold below which operation is
             switched to the alternative model.
 
-    Raises:
-        ValueError: If the class already contains an attribute with the same name
-            as the temporary attribute to be added.
-
     Returns:
         The modified class.
     """
-    # Name of the attribute added to store the alternative model
-    injected_model_attr_name = "_constant_target_model"
-
-    if injected_model_attr_name in (attr.name for attr in cls.__attrs_attrs__):
-        raise ValueError(
-            f"Cannot apply '{catch_constant_targets.__name__}' because "
-            f"'{cls.__name__}' already has an attribute '{injected_model_attr_name}' "
-            f"defined."
-        )
-
     from baybe.surrogates.naive import MeanPredictionSurrogate
 
     # References to original methods
     _fit_original = cls._fit
     _posterior_original = cls._posterior
 
-    def _posterior_new(self, candidates: Tensor) -> tuple[Tensor, Tensor]:
+    def _posterior_new(self, candidates: Tensor) -> Posterior:
         # Alternative model fallback
-        if hasattr(self, injected_model_attr_name):
-            return getattr(self, injected_model_attr_name)._posterior(candidates)
+        if constant_target_model := _constant_target_model_store.get(id(self), None):
+            return constant_target_model._posterior(candidates)
 
         # Regular operation
         return _posterior_original(self, candidates)
 
-    def _fit_new(
-        self, searchspace: SearchSpace, train_x: Tensor, train_y: Tensor
-    ) -> None:
+    def _fit_new(self, train_x: Tensor, train_y: Tensor, context: Any) -> None:
```

Comment:
Docstrong or at least comment what this function does would be appreciated :)

Reply:
added a short one line to each function 👍🏼